### PR TITLE
fix: introduce dom.isHiddenWithCSS for use in dom.isFocusable

### PR DESCRIPTION
**Pull request by jkodu**
This PR does the below:
- creates a new function `isCssVisible` to satisfy requirements from issue https://github.com/dequelabs/axe-core/issues/1208
- swaps the `dom.isVisible` usage in focus computation with `dom.isCssVisible`
- these changes are necessary for completing the rule `aria-hidden-focus` - https://github.com/dequelabs/axe-core/pull/1166, which exists as a WIP PR.

Closes issue:
- https://github.com/dequelabs/axe-core/issues/1208

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [X] Follows the commit message policy, appropriate for next version
- [X] Has documentation updated, a DU ticket, or requires no documentation change
- [X] Includes new tests, or was unnecessary
- [X] Code is reviewed for security by: << Marcy Sutton >>

### CODE REVIEW

File: lib/commons/dom/is-css-visible.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	const nodeName = el.nodeName.toUpperCase();
+	if (['STYLE', 'SCRIPT'].includes(nodeName)) {
+		return false;
+	}
+
+	const style = window.getComputedStyle(el, null);
+	if (!style) {
+		return false;
```
[the remark this answers is not in the document]
Comment (jkodu) [reply lib/commons/dom/is-css-visible.js]: I think this may be a bit intrusive to `throw` for this reason. Happy to discuss.

**Comment by jkodu** [comment]
@WilcoFiers - ready for review.

File: lib/commons/dom/is-hidden-with-css.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		el = el.host; // swap to host node
+	}
+
+	if (['STYLE', 'SCRIPT'].includes(el.nodeName.toUpperCase())) {
+		return false;
+	}
+
+	const style = window.getComputedStyle(el, null);
+	if (!style) {
+		throw new Error('Style does not exist for the given element.');
```
Comment (marcysutton) [inline lib/commons/dom/is-hidden-with-css.js]: Where would this scenario occur?

File: lib/commons/dom/is-hidden-with-css.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Determine whether an element is hidden based on css
+ * @method isHiddenWithCSS
+ * @memberof axe.commons.dom
+ * @instance
+ * @param {HTMLElement} el The HTML Element
+ * @param {Boolean} descendentVisibilityValue (Optional) immediate descendant visibility value used for recursive computation
+ * @return {Boolean} the element's hidden status
+ */
+dom.isHiddenWithCSS = function isHiddenWithCSS(el, descendentVisibilityValue) {
```
Comment (marcysutton) [inline lib/commons/dom/is-hidden-with-css.js]: I like the new name for this method. 👍 

Are you pulling any of the code out of is-visible, or is that mostly duplicate at this point?

File: test/commons/dom/is-hidden-with-css.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+	it('should return false for off-screen fixed positioned element', function() {
+		fixture.innerHTML =
+			'<div id="target" style="position: fixed; top: -9999px">Hi</div>';
+		var el = document.getElementById('target');
+		var actual = isHiddenWithCSSFn(el);
+		assert.isFalse(actual);
+	});
+
+	it('should return false on detached elements', function() {
```
Comment (marcysutton) [inline test/commons/dom/is-hidden-with-css.js]: When would this case arise? Would these even be returned from querying the DOM?

File: lib/commons/dom/is-hidden-with-css.js (hunk 2 above)
Comment (WilcoFiers) [reply lib/commons/dom/is-hidden-with-css.js]: I think this happens with unmounted elements.

File: lib/commons/dom/is-hidden-with-css.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+	if (style.getPropertyValue('display') === 'none') {
+		return true;
+	}
+
+	const visibilityValue = style.getPropertyValue('visibility');
+	if (
+		(visibilityValue === 'hidden' && !descendentVisibilityValue) ||
+		(visibilityValue === 'hidden' &&
+			(descendentVisibilityValue && descendentVisibilityValue !== 'visible'))
```
Comment (WilcoFiers) [inline lib/commons/dom/is-hidden-with-css.js]: should be `=== 'hidden'` There are more values on CSS visibility than just hidden and visible.

File: lib/commons/dom/is-hidden-with-css.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	) {
+		return true;
+	}
+
+	const parent = dom.getComposedParent(el);
+	if (parent) {
+		return dom.isHiddenWithCSS(parent, visibilityValue);
```
Comment (WilcoFiers) [inline lib/commons/dom/is-hidden-with-css.js]: Nice try on the recursion, but it doesn't work this way. This function shows `#foo` as hidden, which it isn't:

```html
<div style="visibility:hidden">
  <div style="visibility:hidden">
    <div style="visibility:visible" id="foo">
      Visible
    </div>
  </div>
</div>
```

File: lib/commons/dom/is-hidden-with-css.js (hunk 3 above)
Comment (jkodu) [reply lib/commons/dom/is-hidden-with-css.js]: This has some replication from `is-visible` and `is-hidden`. We may need to do some refactor afterwards.

File: test/commons/dom/is-hidden-with-css.js (hunk 4 above)
Comment (jkodu) [reply test/commons/dom/is-hidden-with-css.js]: You are right, probably won't be returned when querying the DOM. No harm in this test case, as that is exactly what we are testing.

File: lib/commons/dom/is-hidden-with-css.js (hunk 6 above)
Comment (jkodu) [reply lib/commons/dom/is-hidden-with-css.js]: Have tackled for the above.

File: lib/commons/dom/is-hidden-with-css.js (hunk 2 above)
Comment (marcysutton) [reply lib/commons/dom/is-hidden-with-css.js]: I don't even know how that would come up in the context of an audit, but ok :)

**Pull request merged**